polkadot/runtime/parachains/src/paras/mod.rs

```diff
@@ -1621,43 +1612,45 @@ impl<T: Config> Pallet<T> {
 		now: BlockNumberFor<T>,
 		relay_parent_number: BlockNumberFor<T>,
 		cfg: &configuration::HostConfiguration<BlockNumberFor<T>>,
-		set_go_ahead: SetGoAhead,
+		enact_upgrade_directly: EnactUpgradeDirectly,
 	) -> Weight {
 		let mut weight = Weight::zero();
 
-		// Compute the relay-chain block number starting at which the code upgrade is ready to be
-		// applied.
-		//
-		// The first parablock that has a relay-parent higher or at the same height of `expected_at`
-		// will trigger the code upgrade. The parablock that comes after that will be validated
-		// against the new validation code.
-		//
-		// Here we are trying to choose the block number that will have `validation_upgrade_delay`
-		// blocks from the relay-parent of inclusion of the the block that scheduled code upgrade
-		// but no less than `minimum_validation_upgrade_delay`. We want this delay out of caution
-		// so that when the last vote for pre-checking comes the parachain will have some time until
-		// the upgrade finally takes place.
-		let expected_at = cmp::max(
-			relay_parent_number + cfg.validation_upgrade_delay,
-			now + cfg.minimum_validation_upgrade_delay,
-		);
+		if enact_upgrade_directly == EnactUpgradeDirectly::Yes {
+			FutureCodeHash::<T>::remove(&id);
+			weight.saturating_add(Self::set_current_code(id, *code_hash, now));
+		} else {
+			// Compute the relay-chain block number starting at which the code upgrade is ready to
+			// be applied.
+			//
+			// The first parablock that has a relay-parent higher or at the same height of
+			// `expected_at` will trigger the code upgrade. The parablock that comes after that will
+			// be validated against the new validation code.
+			//
+			// Here we are trying to choose the block number that will have
+			// `validation_upgrade_delay` blocks from the relay-parent of inclusion of the the block
+			// that scheduled code upgrade but no less than `minimum_validation_upgrade_delay`. We
+			// want this delay out of caution so that when the last vote for pre-checking comes the
+			// parachain will have some time until the upgrade finally takes place.
```

Comment:
note that we pick the maximum here - the validation_upgrade_delay needs to be maintained. Reasoning see [here](https://github.com/paritytech/polkadot-sdk/blob/69434d9a32af6103d7cbde9798cf594027fb620e/polkadot/runtime/parachains/src/configuration.rs#L126). It seems this gets violated by these changes.